### PR TITLE
[pt2] get backward for `multi_margin_loss` via `DispatchKey.Autograd`

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #104846
* #104844
* #104578

### CODE REVIEW

**Comment by nkaretnikov** [comment]
Was requested in https://github.com/pytorch/pytorch/pull/104578. Might merge this into the original PR after doing benchmarking/testing.

File: test/expect/HasDecompTest.test_has_decomposition.expect [hunk 1]
```diff
@@ -949,6 +949,8 @@ aten::mps_convolution_backward
 aten::mps_convolution_backward.out
 aten::mps_convolution_transpose_backward
 aten::mps_convolution_transpose_backward.out
+aten::multi_margin_loss_backward
+aten::multi_margin_loss_backward.grad_input
```
Comment (nkaretnikov) [inline test/expect/HasDecompTest.test_has_decomposition.expect:953]: The following test fails when backward is generated via `py_impl`:

```
pytest -rfxs test/test_decomp.py -k test_has_decomposition -v
```

All other tests should pass.

Comment (lezcano) [reply test/expect/HasDecompTest.test_has_decomposition.expect:953]: cc @peterbell10

Comment (lezcano) [reply test/expect/HasDecompTest.test_has_decomposition.expect:953]: Actually, on a second thought, this is expected, so never mind.

**Comment by nkaretnikov** [comment]
I'm going to merge this into the multi_margin_loss PR because inductor fails to compile my handwritten backward. I feel like it's not worth it to debug this because the handwritten decomp will be replaced with an autogen'd one anyway.

```
decomp torch.Size([10, 10]), target=torch.Size([10]), weight=torch.Size([10])

def decomp(input, target, weight):
    def f(input, target, weight):
        return torch.ops.aten.multi_margin_loss(input, target=target, weight=weight)
    f_opt = torch.compile(f, dynamic=True)
    return f_opt(input, target, weight)

torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
AttributeError: 'SymFloat' object has no attribute 'shape'
```

**Pull request closed without merging**